core/src/com/unciv/logic/map/MapGenerator.kt

```diff
@@ -371,15 +372,66 @@ class MapGenerator(val ruleset: Ruleset) {
 
     // Here, we need each specific resource to be spread over the map - it matters less if specific resources are near each other
     private fun spreadStrategicResources(tileMap: TileMap, distance: Int) {
-        val strategicResources = ruleset.tileResources.values.filter { it.resourceType == ResourceType.Strategic }
-        // passable land tiles (no mountains, no wonders) without resources yet
-        val candidateTiles = tileMap.values.filter { it.resource == null && it.isLand && !it.getLastTerrain().impassable }
-        val totalNumberOfResources = candidateTiles.size * tileMap.mapParameters.resourceRichness
-        val resourcesPerType = (totalNumberOfResources/strategicResources.size).toInt()
+        // Extensive but fast preparations - not many terrains therefore all collections are very small
+        // What, including terrain features, can count as land or water, respectively?
+        // This would make little sense with the vanilla ruleset but with this we allow some pretty free modding
+        val baseLand = ruleset.terrains.values.asSequence().filter { it.type == TerrainType.Land }
+        val landFeatures = ruleset.terrains.values.asSequence().filter { it.type == TerrainType.TerrainFeature && it.occursOn != null && it.occursOn.any { it in baseLand.map { it.name } } }
+        val allLand = (baseLand+landFeatures).map { it.name }.toHashSet()
+        val baseWater = ruleset.terrains.values.asSequence().filter { it.type == TerrainType.Water }
+        val waterFeatures = ruleset.terrains.values.asSequence().filter { it.type == TerrainType.TerrainFeature && it.occursOn != null && it.occursOn.any { it in baseWater.map { it.name } } }
+        val allWater = (baseWater+waterFeatures).map { it.name }.toHashSet()
+
+        // Now, per resource: Can they occur on land and/or water?
+        data class LandOrWater(val land: Boolean = false, val water: Boolean = false)
+        fun getLandOrWater(resource: TileResource): LandOrWater {
+            return LandOrWater(resource.terrainsCanBeFoundOn.any { it in allLand }, resource.terrainsCanBeFoundOn.any { it in allWater} )
+        }
+        val strategicResources =
+                ruleset.tileResources.values
+                .filter { it.resourceType == ResourceType.Strategic }
+        val landOrWaterMap =
+                strategicResources
+                .map { Pair<String,LandOrWater>(it.name, getLandOrWater(it)) }
+                .toMap()
+
+        // Get all base terrain any strategic resource can occur on / all features
+        // (could use only allOccursOn, benefit is small, but every little bit counts to speed up filtering later)
+        val allOccursOn = strategicResources.flatMap { it.terrainsCanBeFoundOn }.toHashSet()
```

Comment:
Since we don't use this later, we can run the `.map { ruleset.terrains[it]!! }` here - that way we don't do it twice, once for base and once for features

Reply:
Could. Preferred not to so all three are same type and can easier be seen in the debugger that they contain what's expected. Or maybe I couldn't come up with a good name for the val to differentiate that it's a list of Terrain objects instead of a string map.